### ISSUE TITLE
beta.6: TypeError: viewFactory_<name>0 is not a function

### ISSUE THREAD

**Issue by tycho01** [angular/angular#7037]
I received `TypeError: viewFactory_<name>0 is not a function` under certain conditions involving mutually recursive components; plunker [here](http://plnkr.co/edit/3YRGZm). A similar error (triggered without recursion) was discussed in [6404](https://github.com/angular/angular/issues/6404)/[6489](https://github.com/angular/angular/issues/6489), although that was allegedly fixed again after `beta.1`; I'm still getting this on `beta.6`.

**Comment by tycho01** [comment]
So the crash triggers at a statement `viewFactory_MyComp0(renderer, viewManager, app8_ValueComp, [], null, null, null);` dynamically created at `CodeGenViewFactory`'s [`createAndSetComponentView`](https://github.com/angular/angular/blob/0ae77753f34862a6f584e131af1630dbd588b5de/modules/angular2/src/compiler/view_compiler.ts#L240), thrown into a larger expression at [`createViewFactory`](https://github.com/angular/angular/blob/0ae77753f34862a6f584e131af1630dbd588b5de/modules/angular2/src/compiler/view_compiler.ts#L307) which gets called by `TemplateCompiler`'s [`_createViewFactoryCodeGen`](https://github.com/angular/angular/blob/b86829f4927b7bf44acef72a5d0d44975bfaa903/modules/angular2/src/compiler/template_compiler.ts#L294). It appears that upon execution, this `viewFactory_MyComp0` is `null`.

Having little knowledge to debug this since dynamic code generation complicates things, I tried just adding a null check in the generated statement there and [here](https://github.com/angular/angular/blob/42231f571947fbd2db5ea5692d3990470c019990/modules/angular2/src/core/change_detection/change_detection_jit_generator.ts#L497), but without the viewFactory the parts won't render, so seems it's needed anyway (which... makes sense).

In that [`_createViewFactoryCodeGen`](https://github.com/angular/angular/blob/b86829f4927b7bf44acef72a5d0d44975bfaa903/modules/angular2/src/compiler/template_compiler.ts#L294) I'm seeing a pass for `MyComp0` (check from `compMeta.type.name`) with a `viewFactoryExpr` defining a function `viewFactory_MyComp0`; after, I see another pass for `MyTestComp` (the test component containing `MyComp`) which contains the call to `viewFactory_MyComp0` (as `#MODULE[.template.js]viewFactory_MyComp0`) that it would crash on. So apparently that ends up `undefined` even though it should have been, but I'm not sure how to further trace this.

I know this isn't much, but perhaps after solving this the reverted [compiler refactoring](https://github.com/angular/angular/pull/5993) would still turn out viable too.

**Comment by andrehaveman** [comment]
I get a similar error  (in beta 6) `TypeError: viewFactory_Tree0 is not a function` when trying to nest two components inside each other. I'm trying to build a tree component that nests tree items, which again use the tree component. See this [plunker](http://plnkr.co/edit/gAGm8H).

```js
import {Component, Input, forwardRef} from 'angular2/core';
import {NgFor, NgIf} from 'angular2/common';

@Component({
    selector: 'tree',
    directives: [ NgFor, forwardRef(() => TreeItem) ],
    template: `
        <tree-item *ngFor="#item of data" [item]="item">{{item.title}}</tree-item>
    `
})
class Tree {
    @Input() data: any;
}

@Component({
    selector: 'tree-item',
    directives: [ NgIf, Tree ],
    template: `
        <span>{{item.title}}</span>
        <tree *ngIf="item.children && item.children.length" [data]="item.children"></tree>
    `
})
class TreeItem {
    @Input() item: any;
}
```

**Comment by tonimora74** [comment]
I have the same problem. Is there no solution at this time?

**Comment by tycho01** [comment]
I tried to further debug by injecting `console.log` statements into `createAndSetComponentView` -- if this `viewFactory` isn't declared, while it should be, I figured I wanted to know what is. I examined `this` and `window` in that context, but it's not in those. Directly accessing those viewFactories by name does help. I found that each pass for a given component had viewFactories declared for the component in question as well as for parent components. I was able to log the following:

```
------------------------------------
'about to call: viewFactory_Root0'
'viewFactory_Root0', function ...
'viewFactory_A0', null
'viewFactory_B0', null
'viewFactory_C0', null
'viewFactory_D0', null
'viewFactory_E0', null
------------------------------------
'about to call: viewFactory_A0'
'viewFactory_Root0', function ...
'viewFactory_A0', function ...
'viewFactory_B0', null
'viewFactory_C0', null
'viewFactory_D0', null
'viewFactory_E0', null
------------------------------------
'about to call: viewFactory_B0'
'viewFactory_Root0', null
'viewFactory_A0', function ...
'viewFactory_B0', function ...
'viewFactory_C0', function ...
'viewFactory_D0', null
'viewFactory_E0', null
------------------------------------
'about to call: viewFactory_C0'
'viewFactory_Root0', null
'viewFactory_A0', function ...
'viewFactory_B0', function ...
'viewFactory_C0', function ...
'viewFactory_D0', null
'viewFactory_E0', null
------------------------------------
'about to call: viewFactory_D0'
'viewFactory_Root0', null
'viewFactory_A0', null
'viewFactory_B0', null
'viewFactory_C0', function ...
'viewFactory_D0', function ...
'viewFactory_E0', null
------------------------------------
'about to call: viewFactory_E0'
'viewFactory_Root0', null
'viewFactory_A0', null
'viewFactory_B0', null
'viewFactory_C0', null
'viewFactory_D0', function ...
'viewFactory_E0', function ...
------------------------------------
'about to call: viewFactory_C0'
'viewFactory_Root0', null
'viewFactory_A0', null
'viewFactory_B0', null
'viewFactory_C0', null
'viewFactory_D0', null
'viewFactory_E0', function ...
```

From this, I think the problem is clear. View factories get initialized when needed, stay in scope while deemed still needed, then get ditched. However, its judgement here turns out incorrect; it stumbles back upon a previous component through another route in the mutual recursion, while upon this second pass for the component in question the corresponding view factory (here `viewFactory_C0`) has already gone out of scope.

As a potential fix I would propose potentially keeping the view factories around until later, when we could be sure they have outlived their use.

I'm not very familiar with how it works though, particularly w.r.t. how/when view factories go out of scope, so I think we'll need some expert feedback here. It seems @tbosch single-handedly authored the view compiler, so he'd probably be better equipped to answer that.

**Comment by tycho01** [comment]
Checking further, I tried to see if the being in/out of scope of the view factories related to the different caches part of `TemplateCompiler`, but this isn't the case. I'm slowly realizing hacky workarounds involving "if view compiler not defined then let me reinsert its definition for you" don't really work, since due to the recursion things end up in a stackoverflow (infinite loop). I don't know enough to see the real resolution.

I did finally find a workaround however, obvious in retrospect: disabling JIT compilation, so as to compile the views in run-time fashion rather than the code-gen way. This can be achieved with the following imports:

```
import { provide } from 'angular2/core';
import { ChangeDetectorGenConfig } from 'angular2/src/core/change_detection/change_detection';
```

... for the application itself:

```
bootstrap(App, [
  provide(ChangeDetectorGenConfig, {useValue: new ChangeDetectorGenConfig(false, false, false)}),
]);
```

... or for tests:

```
  beforeEachProviders(() => [
    provide(ChangeDetectorGenConfig, {useValue: new ChangeDetectorGenConfig(false, false, false)}),
  ]);
```

(I'm open to suggestions on how to inject a provider for a single test instead.)

... which doesn't solve the issue with JIT enabled, but would at least make a real fix less urgent.

**Comment by gaiottino** [comment]
I got this bug as well when using recursive components and it's still an issue using the latest `beta.11`. Here's an updated plunkr: http://plnkr.co/edit/EinFLuMs9QGs1gDZnYRn?p=preview

Disabling JIT in the `ChangeDetectorGenConfig` still "solves" the problem but I'd rather not. Any word on an official fix or plan for this?

**Comment by tbosch** [comment]
This will be fixed once my pr https://github.com/angular/angular/pull/6567 lands.
See http://plnkr.co/edit/Jmi4faunpt7v2natZMfN?p=preview

**Comment by xealot** [comment]
At the risk if being a :+1:  I would like to throw my plunkr on this :bug: as well.
https://plnkr.co/edit/GkbaNDn2ggRUI3ZofPIg

Pretty simple example using beta 13. Looking forward to this fix!

**Comment by DrewynV** [comment]
__________________________________

Hi! We just wanted to let you know we receievd your email. Thank you for submitting your request. Our aim is to get back to you within 24 hours at the latest. If you have any additional information to add to your request, please reply to this email.


Also please visit our knowledgebase (http://www.support.magicdust.com.au) to browse for further info, we are adding to it all the time.

Or here is our industry report about the 11 Essentials for a Successful Website (http://www.magicdust.com.au/web-design/.

We look forward to helping you succeed online.

Thanks

The Team @ Magicdust Digital

http://support.magicdust.com.au/

Ph (02) 9979 7474
M-F | 9 am-5:00 pm AEST (AEDT in Summer) excluding Public Holidays

------------------------------------------------------------------------------------------------------------
This message was sent to  reply@reply.github.com in reference to Case #166439.
------------------------------------------------------------------------------------------------------------


[[c35959f90f5ac2343d79ec584ee435b07d160bcf-673073275]]

**Comment by tycho01** [comment]
Fixed by tbosch. Thank you.
